modules/apps/transfer/types/denom.go

```diff
@@ -0,0 +1,108 @@
+package types
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"sort"
+	"strings"
+
+	errorsmod "cosmossdk.io/errors"
+
+	cmtbytes "github.com/cometbft/cometbft/libs/bytes"
+)
+
+// Validate performs a basic validation of the Denom fields.
+func (d Denom) Validate() error {
+	// NOTE: base denom validation cannot be performed as each chain may define
+	// its own base denom validation
+	if strings.TrimSpace(d.Base) == "" {
+		return fmt.Errorf("base denomination cannot be blank")
+	}
+
+	if len(d.Trace) != 0 {
+		trace := strings.Join(d.Trace, "/")
+		identifiers := strings.Split(trace, "/")
+
+		if err := validateTraceIdentifiers(identifiers); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Hash returns the hex bytes of the SHA256 hash of the Denom fields using the following formula:
+//
+// hash = sha256(tracePath + "/" + baseDenom)
+func (d Denom) Hash() cmtbytes.HexBytes {
+	hash := sha256.Sum256([]byte(d.GetFullPath()))
+	return hash[:]
+}
+
+// IBCDenom a coin denomination for an ICS20 fungible token in the format
+// 'ibc/{hash(tracePath + baseDenom)}'. If the trace is empty, it will return the base denomination.
+func (d Denom) IBCDenom() string {
```

Comment:
@crodriguezvega suggest `OnChainDenom`, I guess it'd be `denom.OnChain()`. I like the naming direction, but feels it could be a little more descriptive. Any ideas?